Merge NOAA-GSL gsl/develop into main (update GF CU, MYNN SFC/PBL, RUC LSM, bug fixes in UGWP)

climbfuji opened the pull request:
This PR contains
- a metadata bug fix and a bug fix for a sign error in UGWP - this is urgently needed in the authoritative repositories
- major updates to GF convection, including a small change for the radiation coupling when GF convection is used with GFDDL-MP
- updates to RUC LSM, MYNN SFC, MYNN PBL
- updates to `GFS_debug.F90` to include newly added variables and recently removed variables

Associated PRs:

https://github.com/NOAA-EMC/fv3atm/pull/435
https://github.com/ufs-community/ufs-weather-model/pull/937

For testing, see https://github.com/ufs-community/ufs-weather-model/pull/937

grantfirl commented on physics/GFS_debug.F90:
```diff
@@ -1301,18 +1306,12 @@ subroutine GFS_interstitialtoscreen_run (Model, Statein, Stateout, Sfcprop, Coup
                      call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Interstitial%scmpsw%nirdf        ', Interstitial%scmpsw%nirdf            )
                      call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Interstitial%scmpsw%visbm        ', Interstitial%scmpsw%visbm            )
                      call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Interstitial%scmpsw%visdf        ', Interstitial%scmpsw%visdf            )
-                     call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Interstitial%semis_ice           ', Interstitial%semis_ice               )
```
Some of these changes already exist in #787 , but it's not a big deal since it will be merged first.

climbfuji replied:
Yes, there will be merge conflicts, I am aware of it. I needed those in there to use the GFS_debug code in gsl/develop.

grantfirl commented on physics/GFS_rrtmg_pre.F90:
```diff
@@ -725,13 +726,25 @@ subroutine GFS_rrtmg_pre_run (im, levs, lm, lmk, lmp, n_var_lndp,        &
                 endif
               enddo
             enddo
+          elseif (imfdeepcnv == imfdeepcnv_gf .and. kdt>1) THEN
```
I'm confused by this logic. Can't this condition and the condition above (do_mynnedmf) both be true simultaneously? If so, I'm not sure an IF/ELSEIF is the best way to handle this. Am I missing something?

hannahcbarnes replied:
If MYNN is present , then it takes care of everything (ie: it acounts for all subgrid clouds in MYNN and GF). If GF is called and MYNN is not present, the subgrid clouds in GF must be account for separately (since they are not accounted for in GFDL microphysics) This is why the elseif imfdeepcnv_gf is present and why it is listed after the if MYNN statement. 

If you can come up with better logic I am open

grantfirl replied:
OK, so I definitely was missing some information. Your solution obviously works for the use case, I'm just wondering if there is a clearer way to write it (that won't be interpreted as a bug by subsequent folks reading the code). What about nesting IF's?
`if ((imfdeepcnv == imfdeepcnv_gf .or. do_mynnedmf) .and. kdt>1) then`
   `if (do_mynnedmf) then`
     `!calculate cldcov for MYNN`
   `else`
     `!calculate cldcov for GF`
   `endif`
`else`
  `!rest of options`
`endif`

I don't know if this is any clearer or not. It's less efficient since you're testing for do_mynnedmf twice, but that is probably insignificant. @climbfuji opinions? We could always leave it as-is and maybe putting in a comment explaining things would be sufficient.

hannahcbarnes replied:
I agree, the logic in this part of the code is weird and should be made more clear. I am good with your suggested code change.

grantfirl commented on physics/cu_gf_deep.F90:
```diff
@@ -2410,18 +2448,23 @@ subroutine cup_dd_edt(ierr,us,vs,z,ktop,kbcon,edt,p,pwav, &
-!              prop_c=.9/aeroadd
+               aeroadd=0.
+               if((psumh(i)>0.).and.(psum2(i)>0.))then
+               aeroadd=((1.e-2*ccnclean)**beta3)*((psumh(i)*1.e0)**(alpha3-1))
```
Why are we multiplying by 1e.0 in this line and 2459? Is this a placeholder to be changed in the future or something?

hannahcbarnes replied:
You can remove the 1e.0

grantfirl commented on physics/cu_gf_driver.meta:
```diff
@@ -301,6 +322,14 @@
   kind = kind_phys
   intent = in
+[aod_gf]
+  standard_name = aod_gf_deep
```
We should use "aerosol_optical_depth" rather than "aod". The standard names aerosol_optical_depth_for_shortwave_bands_01_16 and  aerosol_optical_depth_for_longwave_bands_01_16 already exist. How does this new variable relate? Is it for all radiation bands? Is there a reason that this is specific to GF (could it potentially be used in other schemes)?

hannahcbarnes replied:
This is a specific variable used for only GF. It is not related to aerosol_optical_depth_for_shortwave_bands_01_16 or aerosol_optical_depth_for_longwave_bands_01_16. It is not for radiation bands, it is used as a measure of pollution in the atmosphere. GF is aerosol-aware, which is why it is needed. SAS is also aerosol-aware, but I have not looked close enough into the details of that scheme to see if they would use this. 

You could rename it as aerosol_optical_depth_gf.  You can also provide additional suggestions if you wish.

grantfirl replied:
I'm thinking that we can just use aerosol_optical_depth full stop. According to the standard name rules ([link](https://github.com/ESCOMP/CCPPStandardNames/blob/main/StandardNamesRules.rst)), we shouldn't add qualifiers like _gf if they aren't absolutely necessary. That way if, say, SAS or any other convective scheme wanted to reuse this variable for their purposes or the same purpose, there wouldn't necessarily be a need to add a new variable (and therefore the memory footprint stays the same).

hannahcbarnes replied:
I am worried about that since people may get confused with aerosol optical depth used in other parts of the code, such as in the radiation. This variable should absolutely not be used in radiation. The name needs to be at least somewhat specific.

climbfuji replied:
I agree.

grantfirl replied:
@climbfuji agree with what? The need to qualify or trying to keep it general? If it shouldn't be used for radiation, we could qualify that it is to be used in convection schemes. That would at least be more general than saying that only GF can use it?

climbfuji replied:
I agree with Hannah that it needs to be made clear that this is specific to GF (I don't even know if it is compatible with any other convection scheme, but if it is, then it would be better to express it in the standard name). We do have other such cases in CCPP, of course, where we have adopted standard names w/o qualifier. One example being the water/ice friendly aerosols that, in the current representation can only be used by Thompson MP. Of course, if Thompson is not used, than anybody else can use those variables the way they want, but a developer finding aerosols somewhere and thinking that he or she can just use it will run into trouble.

grantfirl replied:
OK, you both have convinced me. :) To be consistent with existing standard names spelling out GF, how about let's settle on aerosol_optical_depth_for_grell_freitas_deep_convection? Definitely longer, but consistent with others.

climbfuji replied:
Heureka! Good with me. I'll make the change along with others over the weekend.

hannahcbarnes replied:
Thanks, this good with me as well.

climbfuji commented:
@hannahcbarnes @haiqinli @joeolson42 Please take note of the comments that @grantfirl left and let me know if you have any suggestions/changes you wish to make. Thanks!

grantfirl commented on physics/drag_suite.meta:
```diff
@@ -497,7 +497,7 @@
   long_name = multiplic. factors for (1) mountain blocking drag coeff. and (2) ref. level orographic gravity wave drag
   units = none
-  dimensions = (2)
+  dimensions = (4)
```
Should we define a variable for this dimension rather than hardcode it? @climbfuji do you remember if there is a CCPP metadata rule to this effect (not to use hardcoded dimensions)?

climbfuji replied:
Integers are allowed, yes, and so are constants (e.g. there is already a `ccpp_one` or something like that). Creating a new dimension for it is possible, of course, whether it is necessary is a different question.

grantfirl replied:
OK, it's fine to leave it as-is then. The only reason I asked is that a change was made in the metadata but not in the corresponding Fortran in this PR, which points to the problem of hard-coding in multiple places (leads to potential bugs).

climbfuji replied:
The dimension in the Fortran file should be `(:)` and no fixed dimension. Fortunately, with the new debugging capability (check dimensions in auto-generated caps in debug mode) and the use of assumed-shape arrays, any such inconsistency will be caught.

grantfirl replied:
Ah yes, I forgot about the insistence on (:). All good!

grantfirl commented on physics/module_MYNNPBL_wrapper.F90:
```diff
@@ -222,8 +221,9 @@ SUBROUTINE mynnedmf_wrapper_run(        &
-     &       levflag=2
+     &       bl_mynn_mixscalars=1
+      REAL, PARAMETER ::                                    &
+     &       closure=2.6   !2.5, 2.6 or 3.0
```
Would there be any benefit of exposing this to the physics namelist, or do you not want folks messing with it?

climbfuji replied:
I think not. I don't think it makes sense to expose every tuning parameter as a control when only experts know what it is and how to set it. Pretty much the same what Greg said the other day about exposing tuning knobs in Thompson MP.

grantfirl replied:
OK. I'm not suggesting exposing every tuning parameter, but this one is a BIG one which (should) determine how turbulent moments are calculated (diagnostic/prognostic). I don't have a problem leaving it hidden if researchers should not be touching it, of course.

climbfuji replied:
@hannahcbarnes what is your take?

hannahcbarnes replied:
This is a better question for @joeolson42

grantfirl replied:
It's also not super-important. It can be exposed to the namelist easily later if ever needed for research purposes.

joeolson42 replied:
I do intend to make this a namelist option in the near future. I still consider it new enough that I prefer it be "protected" from the public for now.

grantfirl commented on physics/module_bl_mynn.F90:
```diff
@@ -490,7 +519,7 @@ SUBROUTINE  mym_initialize (                                &
     REAL :: phm,vkz,elq,elv,b1l,b2l,pmz=1.,phh=1.,flt=0.,flq=0.,tmpq
     REAL :: zi
-    REAL, DIMENSION(kts:kte) :: theta
+      REAL, DIMENSION(kts:kte) :: theta,thetav,thlsg,qwsg
```
Silly, but for readability, consider removing the whitespace change so that var declarations align.

joeolson42 replied:
I agree. It's definitely meant to be aligned.  I probably fat-fingered a couple spaces without noticing.

climbfuji replied:
I'll fix it together with some other changes that were requested.

grantfirl reviewed (approved):
I will not pretend to understand all of the code changes represented here, but I've commented on anything unclear that I've detected. Thanks to all the developers of these schemes and nice work!

climbfuji commented:
@grantfirl @hannahcbarnes I addressed the reviewer comments for the gsl/develop update in https://github.com/NCAR/ccpp-physics/pull/791/commits/83ff12942727e78660b38c9d9eb6bbe722fe0081. Please check the changes, especially the change in the `if-then-else` logic in the RRTMG pre code. Thanks!

climbfuji commented:
@yangfanglin This PR contains the bug fix for the sign error in `physics/drag_suite.F90`. It's been merged just now.

yangfanglin commented:
Thanks, Dom. 

Sent from my iPhone

> On Dec 13, 2021, at 7:55 AM, Dom Heinzeller ***@***.***> wrote:
> 
> ﻿
> @yangfanglin This PR contains the bug fix for the sign error in physics/drag_suite.F90. It's been merged just now.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.
